util/src/main/java/org/hyperledger/besu/util/era1/Era1Reader.java

```diff
@@ -0,0 +1,130 @@
+    int slot = 0;
+    while (inputStream.available() > 0) {
+      Era1Type type = Era1Type.getForTypeCode(inputStream.readNBytes(TYPE_LENGTH));
+      int length = (int) convertLittleEndianBytesToLong(inputStream.readNBytes(LENGTH_LENGTH));
+      switch (type) {
+        case VERSION -> {
+          // do nothing
+        }
+        case EMPTY, ACCUMULATOR, TOTAL_DIFFICULTY -> {
+          // skip the bytes that were indicated to be empty
+          // TODO read ACCUMULATOR and TOTAL_DIFFICULTY properly?
+          inputStream.skipNBytes(length);
+        }
+        case COMPRESSED_EXECUTION_BLOCK_HEADER -> {
+          byte[] compressedExecutionBlockHeader = inputStream.readNBytes(length);
+          try (SnappyFramedInputStream decompressionStream =
+              snappyFactory.createFramedInputStream(compressedExecutionBlockHeader)) {
+            listener.handleExecutionBlockHeader(
+                new Era1ExecutionBlockHeader(decompressionStream.readAllBytes(), slot));
+          }
+        }
+        case COMPRESSED_EXECUTION_BLOCK_BODY -> {
+          byte[] compressedExecutionBlock = inputStream.readNBytes(length);
+          try (SnappyFramedInputStream decompressionStream =
+              snappyFactory.createFramedInputStream(compressedExecutionBlock)) {
+            listener.handleExecutionBlockBody(
+                new Era1ExecutionBlockBody(decompressionStream.readAllBytes(), slot));
+          }
+        }
+        case COMPRESSED_EXECUTION_BLOCK_RECEIPTS -> {
+          byte[] compressedReceipts = inputStream.readNBytes(length);
+          try (SnappyFramedInputStream decompressionStream =
+              snappyFactory.createFramedInputStream(compressedReceipts)) {
+            listener.handleExecutionBlockReceipts(
+                new Era1ExecutionBlockReceipts(decompressionStream.readAllBytes(), slot++));
+          }
+        }
+        case BLOCK_INDEX -> {
+          ByteArrayInputStream slotIndexInputStream =
+              new ByteArrayInputStream(inputStream.readNBytes(length));
+          long startingSlot =
+              convertLittleEndianBytesToLong(slotIndexInputStream.readNBytes(STARTING_SLOT_LENGTH));
+          List<Long> indexes = new ArrayList<>();
+          while (slotIndexInputStream.available() > SLOT_INDEX_COUNT_LENGTH) {
+            indexes.add(
+                convertLittleEndianBytesToLong(slotIndexInputStream.readNBytes(SLOT_INDEX_LENGTH)));
+          }
+          long indexCount =
+              convertLittleEndianBytesToLong(
+                  slotIndexInputStream.readNBytes(SLOT_INDEX_COUNT_LENGTH));
+          if (indexCount != indexes.size()) {
+            LOG.warn(
+                "index count does not match number of indexes present for InputStream: {}",
+                inputStream);
+          }
+          listener.handleBlockIndex(new Era1BlockIndex(startingSlot, indexes));
+        }
+      }
+    }
+  }
+
+  private long convertLittleEndianBytesToLong(final byte[] bytes) {
```

Comment:
This should go in EndianUtils and consider using either inline in EndianUtils or directly from bouncy castle https://github.com/bcgit/bc-java/blob/main/core/src/main/java/org/bouncycastle/util/Pack.java#L303 as the bit shift is quite likely more efficient than using a byte buffer